### PR TITLE
solana/program_v2: Add constraints for pass operations ATAs

### PR DESCRIPTION
**Pull request by JulianSpring**
Added constraints to issue_pass, verify_pass, expire_pass and refresh_pass for the atas that are passed in.

### CODE REVIEW

File: solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs [hunk 1]
```diff
@@ -43,6 +43,10 @@ impl Gatekeeper {
             + OC_SIZE_VEC_PREFIX + GatekeeperAuthKey::ON_CHAIN_SIZE * auth_keys
         // auth keys
     }
+    // Checks if the network_token_account is a valid token account for the network
+    pub fn check_network_ownership(&self, gatekeeper: &[Pubkey], owner: Pubkey) -> bool {
+        gatekeeper.iter().any(|&key| key == owner)
```
Comment (rado0x54) [inline solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs:48]: I'm not sure I understand this. The network_token_account is owned by the network and has no relationship to the Gatekeepers.

File: solana/program_v2/programs/gateway_v2/src/instructions/gatekeeper/refresh_pass.rs [hunk 2]
```diff
@@ -62,8 +62,14 @@ pub struct PassRefresh<'info> {
     pub mint_account: Account<'info, Mint>,
     #[account(mut)]
     pub funder_token_account: Account<'info, TokenAccount>,
-    #[account(mut)]
+    #[account(
+    mut,
+    constraint = gatekeeper.check_network_ownership(& network.gatekeepers, network_token_account.owner),
```
Comment (rado0x54) [inline solana/program_v2/programs/gateway_v2/src/instructions/gatekeeper/refresh_pass.rs:67]: I feel you placed these constraints exactly wrong :)

File: solana/program_v2/packages/tests/src/test-set-up.ts [hunk 3]
```diff
@@ -124,10 +124,10 @@ export const setUpAdminNetworkGatekeeper = async (
   await setGatekeeperFlagsAndFees(stakingPDA, networkService, 65535, [
     {
       token: mint,
-      issue: new anchor.BN(1000),
-      refresh: new anchor.BN(1000),
-      expire: new anchor.BN(1000),
-      verify: new anchor.BN(1000),
+      issue: 1000,
```
[the remark this answers is not in the document]
Comment (JulianSpring) [reply solana/program_v2/packages/tests/src/test-set-up.ts:127]: There is a guard for more than 100% `in calculate_network_and_gatekeeper_fee` and a test for 100% I guess I should add one for negative and greater than 100% thanks.

File: solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs (hunk 1 above)
Comment (rado0x54) [reply solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs:48]: I think what you want to check is if `gatekeeper_token_account` belongs to any GATEKEEPER in the network? (is this sufficient)? Don't you KNOW the actual gatekeeper during issue?

Comment (rado0x54) [reply solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs:48]: So I would not expect that you iterate.

File: solana/program_v2/packages/tests/src/test-set-up.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (JulianSpring) [reply solana/program_v2/packages/tests/src/test-set-up.ts:127]: This does not related to fee percentages but to absolute fee values

File: solana/program_v2/packages/client/core/src/GatekeeperService.ts [hunk 4]
```diff
@@ -245,6 +247,7 @@ export class GatekeeperService extends AbstractService {
         networkTokenAccount: networkTokenAccount,
         gatekeeperTokenAccount: gatekeeperTokenAccount,
         funderTokenAccount: funderTokenAccount,
+        feePayer: funderAuthority,
```
Comment (rado0x54) [inline solana/program_v2/packages/client/core/src/GatekeeperService.ts:250]: resolve naming inconsistency. I'd propose rentPayer, but this should then be consistent across all interfaces.

File: solana/program_v2/packages/tests/src/pass-suite/issue.test.ts [hunk 5]
```diff
@@ -75,13 +75,14 @@ describe('issue', () => {
...
-        gatekeeperAta.address,
         networkAta.address,
+        gatekeeperAta.address,
         funderAta.address,
         funderKeypair.publicKey
       )
       .withPartialSigners(funderKeypair)
-      .rpc();
+      .rpc()
+      .catch((e) => console.log(e));
```
Comment (rado0x54) [inline solana/program_v2/packages/tests/src/pass-suite/issue.test.ts:85]: Remove / don't blanket catch errors.

File: solana/program_v2/programs/gateway_v2/src/instructions/gatekeeper/expire_pass.rs [hunk 6]
```diff
@@ -12,7 +12,7 @@ pub fn expire_pass(ctx: Context<PassExpire>) -> Result<()> {
     let pass = &mut ctx.accounts.pass;
     let network = &mut ctx.accounts.network;
     let gatekeeper = &mut ctx.accounts.gatekeeper;
-    let payer = &mut ctx.accounts.payer;
+    let funder = &mut ctx.accounts.fee_payer;
```
Comment (rado0x54) [inline solana/program_v2/programs/gateway_v2/src/instructions/gatekeeper/expire_pass.rs:15]: also here. Consistent naming its funder -> fee_payer -> funder at the moment :)

**Review by rado0x54 — CHANGES_REQUESTED** [review]
Just two small things.

File: solana/program_v2/packages/client/core/src/GatekeeperService.ts (hunk 4 above)
Comment (JulianSpring) [reply solana/program_v2/packages/client/core/src/GatekeeperService.ts:250]: This is the key paying for the passes, think funder makes sense? Since we have a funder token account.

File: solana/program_v2/packages/tests/src/pass-suite/change_pass_gatekeeper.test.ts [hunk 7]
```diff
@@ -70,10 +70,12 @@ describe('Change pass gatekeeper', () => {
         subject.publicKey,
         TOKEN_PROGRAM_ID,
```
Comment (william-brooks) [inline solana/program_v2/packages/tests/src/pass-suite/change_pass_gatekeeper.test.ts:71]: We actually don't need to pass this in as it's always going to be the TOKEN_PROGRAM_ID (i.e. the same token program we check on chain)...

**Review by william-brooks — APPROVED** [review]
